src/main/java/org/takes/rq/RqCookies.java

```diff
@@ -35,86 +35,103 @@
-
-    /**
-     * Ctor.
-     * @param req Original request
-     */
-    public RqCookies(final Request req) {
-        super(req);
-    }
-
+public interface RqCookies extends Request {
     /**
      * Get single cookie.
-     * @param key Cookie name
+     * @param name Cookie name
      * @return List of values (can be empty)
      * @throws IOException If fails
      */
-    public Iterable<String> cookie(final CharSequence key) throws IOException {
-        final Map<String, String> map = this.map();
-        final String name = key.toString().toLowerCase(Locale.ENGLISH);
-        final String value = map.get(name);
-        final Iterable<String> iter;
-        if (value == null) {
-            iter = new VerboseIterable<String>(
-                Collections.<String>emptyList(),
-                new Sprintf(
-                    "there are no Cookies by name \"%s\" among %d others: %s",
-                    key, map.size(), map.keySet()
-                )
-            );
-        } else {
-            iter = new VerboseIterable<String>(
-                Collections.singleton(value),
-                new Sprintf(
-                    "there is always only one Cookie by name \"%s\"",
-                    key
-                )
-            );
-        }
-        return iter;
-    }
+    Iterable<String> cookie(CharSequence name) throws IOException;
 
     /**
      * Get all cookie names.
      * @return All names
      * @throws IOException If fails
      */
-    public Iterable<String> names() throws IOException {
-        return this.map().keySet();
-    }
+    Iterable<String> names() throws IOException;
 
     /**
-     * Parse them all in a map.
-     * @return Map of them
-     * @throws IOException If fails
+     * Request decorator, for HTTP cookies parsing.
+     *
+     * <p>The class is immutable and thread-safe.
+     * @author Dmitry Zaytsev (dmitry.zaytsev@gamil.com)
+     * @version $Id$
+     * @since 0.13.1
      */
-    private Map<String, String> map() throws IOException {
-        final ConcurrentMap<String, String> map =
-            new ConcurrentHashMap<String, String>(0);
-        final Iterable<String> values = new RqHeaders(this).header("Cookie");
-        for (final String value : values) {
-            for (final String pair : value.split(";")) {
-                final String[] parts = pair.split("=", 2);
-                final String key = parts[0].trim().toLowerCase(Locale.ENGLISH);
-                if (parts.length > 1 && !parts[1].isEmpty()) {
-                    map.put(key, parts[1].trim());
-                } else {
-                    map.remove(key);
+    @EqualsAndHashCode(callSuper = true)
+    final class Base extends RqWrap implements RqCookies {
+        /**
+         * Ctor.
+         * @param req Original request
+         */
+        public Base(final Request req) {
+            super(req);
+        }
+
+        @Override
+        public Iterable<String> cookie(final CharSequence key)
```

Comment:
@yegor256 Why cookie(name) should return Iterable? There are ZERO or ONE possible value for one key